File: app/src/main/java/com/kidozh/discuzhub/activities/ui/home/HomeViewModel.java
package com.kidozh.discuzhub.activities.ui.home;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.kidozh.discuzhub.R;
import com.kidozh.discuzhub.entities.ErrorMessage;
import com.kidozh.discuzhub.entities.Discuz;
import com.kidozh.discuzhub.entities.forumUserBriefInfo;
import com.kidozh.discuzhub.results.DiscuzIndexResult;
import com.kidozh.discuzhub.services.DiscuzApiService;
import com.kidozh.discuzhub.utilities.NetworkUtils;

import java.util.List;



import okhttp3.OkHttpClient;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;


public class HomeViewModel extends AndroidViewModel {
    private final static String TAG = HomeViewModel.class.getSimpleName();

    private MutableLiveData<String> mText;
    private MutableLiveData<List<DiscuzIndexResult.ForumCategory>> forumCategories;
    public MutableLiveData<ErrorMessage> errorMessageMutableLiveData = new MutableLiveData<>(null);
    public MutableLiveData<forumUserBriefInfo> userBriefInfoMutableLiveData;
    public MutableLiveData<Boolean> isLoading;
    public MutableLiveData<DiscuzIndexResult> bbsIndexResultMutableLiveData;

    Discuz bbsInfo;
    forumUserBriefInfo userBriefInfo;


    public HomeViewModel(Application application) {
        super(application);
        
        
        isLoading = new MutableLiveData<>(false);
        bbsIndexResultMutableLiveData = new MutableLiveData<>(null);
        userBriefInfoMutableLiveData =  new MutableLiveData<>(null);

    }

    public void setBBSInfo(@NonNull Discuz bbsInfo, forumUserBriefInfo userBriefInfo){
        this.bbsInfo = bbsInfo;
        this.userBriefInfo = userBriefInfo;
        userBriefInfoMutableLiveData = new MutableLiveData<>(userBriefInfo);

    }

    public LiveData<List<DiscuzIndexResult.ForumCategory>> getForumCategoryInfo(){
        if(forumCategories == null){
            forumCategories = new MutableLiveData<List<DiscuzIndexResult.ForumCategory>>();
            loadForumCategoryInfo();
        }
        return forumCategories;
    }

    public void loadForumCategoryInfo(){
        if(!NetworkUtils.isOnline(getApplication())){
            isLoading.postValue(false);
            errorMessageMutableLiveData.postValue(NetworkUtils.getOfflineErrorMessage(getApplication()));
            return;
        }


        OkHttpClient client = NetworkUtils.getPreferredClientWithCookieJarByUser(this.getApplication(),userBriefInfo);
        Retrofit retrofit = NetworkUtils.getRetrofitInstance(bbsInfo.base_url,client);
        DiscuzApiService service = retrofit.create(DiscuzApiService.class);
        Call<DiscuzIndexResult> bbsIndexResultCall = service.indexResult();
        isLoading.postValue(true);

        bbsIndexResultCall.enqueue(new Callback<DiscuzIndexResult>() {
            @Override
            public void onResponse(Call<DiscuzIndexResult> call, Response<DiscuzIndexResult> response) {
                if(response.isSuccessful() && response.body()!=null){
                    DiscuzIndexResult indexResult = response.body();
                    bbsIndexResultMutableLiveData.postValue(indexResult);
                    if(indexResult.forumVariables !=null){
                        forumUserBriefInfo serverReturnedUser = indexResult.forumVariables.getUserBriefInfo();
                        userBriefInfoMutableLiveData.postValue(serverReturnedUser);
                        errorMessageMutableLiveData.postValue(null);
                        // prepare to render index page
                        List<DiscuzIndexResult.ForumCategory> categoryList = indexResult.forumVariables.forumCategoryList;
                        forumCategories.postValue(categoryList);


                    }
                    else {
                        if(indexResult.message !=null){
                            errorMessageMutableLiveData.postValue(indexResult.message.toErrorMessage());

                        }
                        else if(indexResult.error.length()!=0){
                            errorMessageMutableLiveData.postValue(new ErrorMessage(
                                    getApplication().getString(R.string.discuz_api_error),
                                    indexResult.error
                            ));

                        }
                        else {
                            errorMessageMutableLiveData.postValue(new ErrorMessage(
                                    getApplication().getString(R.string.empty_result),
                                    getApplication().getString(R.string.discuz_network_result_null)
                            ));
                        }
                    }
                }
                else {
                    errorMessageMutableLiveData.postValue(new ErrorMessage(String.valueOf(response.code()),
                            getApplication().getString(R.string.discuz_network_unsuccessful,response.message())));
                }
                isLoading.postValue(false);
            }

            @Override
            public void onFailure(Call<DiscuzIndexResult> call, Throwable t) {
                errorMessageMutableLiveData.postValue(new ErrorMessage(
                        getApplication().getString(R.string.discuz_network_failure_template),
                        t.getLocalizedMessage() == null?t.toString():t.getLocalizedMessage()
                ));
                isLoading.postValue(false);
            }
        });
    }

    public LiveData<String> getText() {
        return mText;
    }
}

File: app/src/main/java/com/kidozh/discuzhub/activities/ui/home/HomeViewModel.kt
package com.kidozh.discuzhub.activities.ui.home

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.kidozh.discuzhub.R
import com.kidozh.discuzhub.activities.ui.home.HomeViewModel
import com.kidozh.discuzhub.entities.Discuz
import com.kidozh.discuzhub.entities.ErrorMessage
import com.kidozh.discuzhub.entities.User
import com.kidozh.discuzhub.results.DiscuzIndexResult
import com.kidozh.discuzhub.services.DiscuzApiService
import com.kidozh.discuzhub.utilities.NetworkUtils
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class HomeViewModel(application: Application) : AndroidViewModel(application) {
    private var forumCategories: MutableLiveData<List<DiscuzIndexResult.ForumCategory>>? = null
    var errorMessageMutableLiveData = MutableLiveData<ErrorMessage?>(null)
    var userBriefInfoMutableLiveData: MutableLiveData<User?>
    var isLoading: MutableLiveData<Boolean> = MutableLiveData(false)
    var bbsIndexResultMutableLiveData: MutableLiveData<DiscuzIndexResult?> = MutableLiveData(null)
    lateinit var bbsInfo: Discuz
    var userBriefInfo: User? = null
    fun setBBSInfo(bbsInfo: Discuz, userBriefInfo: User?) {
        this.bbsInfo = bbsInfo
        this.userBriefInfo = userBriefInfo
        userBriefInfoMutableLiveData = MutableLiveData(userBriefInfo)
    }

    val forumCategoryInfo: LiveData<List<DiscuzIndexResult.ForumCategory>>
        get() {
            if (forumCategories == null) {
                forumCategories = MutableLiveData()
                loadForumCategoryInfo()
            }
            return forumCategories!!
        }

    fun loadForumCategoryInfo() {
        if (!NetworkUtils.isOnline(getApplication())) {
            isLoading.postValue(false)
            errorMessageMutableLiveData.postValue(NetworkUtils.getOfflineErrorMessage(getApplication()))
            return
        }
        val client = NetworkUtils.getPreferredClientWithCookieJarByUser(getApplication(), userBriefInfo)



        val retrofit = NetworkUtils.getRetrofitInstance(bbsInfo.base_url, client)
        val service = retrofit.create(DiscuzApiService::class.java)
        val bbsIndexResultCall = service.indexResult()
        isLoading.postValue(true)
        Log.d(TAG,"Send raw request to ${service.indexResultRaw().request()}")

        bbsIndexResultCall.enqueue(object : Callback<DiscuzIndexResult> {
            override fun onResponse(
                call: Call<DiscuzIndexResult>,
                response: Response<DiscuzIndexResult?>
            ) {
                if (response.isSuccessful && response.body() != null) {
                    val indexResult = response.body()!!
                    Log.d(TAG,"GET index response message ${response.message()} user ${userBriefInfo}")
                    bbsIndexResultMutableLiveData.postValue(indexResult)

                    val serverReturnedUser = indexResult.forumVariables.userBriefInfo
                    userBriefInfoMutableLiveData.postValue(serverReturnedUser)
                    errorMessageMutableLiveData.postValue(null)
                    // prepare to render index page
                    val categoryList :List<DiscuzIndexResult.ForumCategory> = indexResult.forumVariables.forumCategoryList
                    Log.d(TAG,"GET category list size ${indexResult.forumVariables.forumCategoryList.size}")
                    forumCategories!!.postValue(categoryList)
                } else {
                    errorMessageMutableLiveData.postValue(
                        ErrorMessage(
                            response.code().toString(),
                            getApplication<Application>().getString(
                                R.string.discuz_network_unsuccessful,
                                response.message()
                            )
                        )
                    )
                }
                isLoading.postValue(false)
            }

            override fun onFailure(call: Call<DiscuzIndexResult?>, t: Throwable) {
                errorMessageMutableLiveData.postValue(
                    ErrorMessage(
                        getApplication<Application>().getString(R.string.discuz_network_failure_template),
                        if (t.localizedMessage == null) t.toString() else t.localizedMessage
                    )
                )
                isLoading.postValue(false)
            }
        })
    }

    companion object {
        private val TAG = HomeViewModel::class.java.simpleName
    }

    init {
        userBriefInfoMutableLiveData = MutableLiveData(null)
    }
}